core/runtime.rs

```diff
@@ -1549,34 +1554,73 @@ impl JsRuntime {
   }
 
   fn drain_macrotasks(&mut self) -> Result<(), Error> {
-    let js_macrotask_cb_handle =
-      match &Self::state(self.v8_isolate()).borrow().js_macrotask_cb {
-        Some(handle) => handle.clone(),
-        None => return Ok(()),
-      };
+    let state = Self::state(self.v8_isolate());
+
+    if state.borrow().js_macrotask_cbs.is_empty() {
+      return Ok(());
+    }
 
+    let js_macrotask_cb_handles = state.borrow().js_macrotask_cbs.clone();
```

Comment:
This also clones all the v8::Global handles in the vec? That's a fairly expensive operation.

Also, I think I'm missing where `js_macrotask_cbs` is emptied. It looks like currently it only grows, never shrinks.

Reply:
Yeah I believe it does clone global handles, in practice there should be no more than two of them, but I can try not to clone them at all (however it's risky because I'd have to hold onto borrow of state and there's nothing preventing callback from using state as well, ergo we might run into `BorrowError`).

> Also, I think I'm missing where js_macrotask_cbs is emptied. It looks like currently it only grows, never shrinks.

Correct, but until this PR there was only a single macrotask callback allowed (used by timers), this PR changes it to allow multiple callbacks. In practice there will be one callback for timers, one for process.nextTick, there might be more of them if users pull multiple versions of `deno_std` (and thus register multiple `process.nextTick` callback.